### PR TITLE
Gameobjects

### PR DESCRIPTION
**Pull request by aaron02**
**Description**
Clean Up Gameobjects, use proper Loot States, Proper Open Closed Handling and Changed Respawn and Despawn Handling
Split Gameobject Spawns Database, Fix some Wrong Positions/Rotations example: coilfang reservior, karazhan, hellife ramparts
fix fishing

deleted 33700 spawn overrides which where 1 to 1 identical to gameobjectproto table.

Wiki: https://github.com/AscEmu/Wiki/pull/46

**Todo / Checklist**
- [x] Target is branch develop.
- [x] Detailed description about this pull request.
- [x] Checked AE-Coding standards.

**Tests Performed:** 
- [x] Build AE.
- [x] Server startup.
- [x] Log into world.

**Multiversion Ingame Tests Performed:**
- [x] WotLK

### CODE REVIEW

**Comment by Zyres** [comment]
Update sql updatefiles.

File: src/world/Objects/GameObject.h [hunk 1]
```diff
@@ -116,6 +160,20 @@ struct GameObjectProperties
...
+            uint32 questList;                   //1
+            uint32 pageMaterial;                //2
+            uint32 gossipID;                    //3
+            uint32 customAnim;                  //4
+            uint32 noDamageImmune;              //5
+            uint32 openTextID;                  //6 can be used to replace castBarCaption?
+            uint32 losOK;                       //7
+            uint32 allowMounted;                //8 Is usable while on mount/vehicle. (0/1)
+            uint32 large;                       //9
+            uint32 conditionID1;                //10
```
Comment (Appled) [inline src/world/Objects/GameObject.h:175]: uint32_t

File: src/world/Objects/GameObject.h [hunk 2]
```diff
@@ -139,6 +197,16 @@ struct GameObjectProperties
...
         // 5 GAMEOBJECT_TYPE_GENERIC
+        struct
+        {
+            uint32 floatingTooltip;             // parameter_0
+            uint32 highlight;                   // parameter_1
+            uint32 serverOnly;                  // parameter_2
+            uint32 large;                       // parameter_3
+            uint32 floatOnWater;                // parameter_4
+            int32 questID;                      // parameter_5
+            uint32 conditionID1;                // parameter_6
```
Comment (Appled) [inline src/world/Objects/GameObject.h:208]: uint32_t int32_t

File: src/world/Objects/GameObject.h [hunk 3]
```diff
@@ -207,6 +275,17 @@ struct GameObjectProperties
...
+        struct
+        {
+            uint32 lockId;                      // parameter_0
+            uint32 radius;                      // parameter_1
+            uint32 damageMin;                   // parameter_2
+            uint32 damageMax;                   // parameter_3
+            uint32 damageSchool;                // parameter_4
+            uint32 autoCloseTime;               // parameter_5 secs till autoclose = autoCloseTime / 0x10000
+            uint32 openTextID;                  // parameter_6
+            uint32 closeTextID;                 // parameter_7
```
Comment (Appled) [inline src/world/Objects/GameObject.h:287]: uint32_t

File: src/world/Objects/GameObject.h [hunk 4]
```diff
@@ -246,6 +325,11 @@ struct GameObjectProperties
         // 19 GAMEOBJECT_TYPE_MAILBOX
         // 20 GAMEOBJECT_TYPE_AUCTIONHOUSE
         // 21 GAMEOBJECT_TYPE_GUARDPOST
+        struct
+        {
+            uint32 creatureID;                              //0
+            uint32 charges;                                 //1
```
Comment (Appled) [inline src/world/Objects/GameObject.h:331]: uint32_t

File: src/world/Objects/GameObject.h [hunk 5]
```diff
@@ -257,6 +341,18 @@ struct GameObjectProperties
...
+        {
+            uint32 lockId;                      // parameter_0
+            uint32 pickupSpell;                 // parameter_1
+            uint32 radius;                      // parameter_2
+            uint32 returnAura;                  // parameter_3
+            uint32 returnSpell;                 // parameter_4
+            uint32 noDamageImmune;              // parameter_5
+            uint32 openTextID;                  // parameter_6
+            uint32 losOK;                       // parameter_7
+            uint32 conditionID1;                // parameter_8
```
Comment (Appled) [inline src/world/Objects/GameObject.h:354]: uint32_t

File: src/world/Objects/GameObject.h [hunk 6]
```diff
@@ -267,9 +363,42 @@ struct GameObjectProperties
             uint32 lock_id;                     // parameter_4 from Lock.dbc
         }fishinghole;
         // 26 GAMEOBJECT_TYPE_FLAGDROP
+        struct
+        {
+            uint32 lockId;                      // parameter_0
+            uint32 eventID;                     // parameter_1
+            uint32 pickupSpell;                 // parameter_2
+            uint32 noDamageImmune;              // parameter_3
+            uint32 openTextID;                  // parameter_4
```
Comment (Appled) [inline src/world/Objects/GameObject.h:372]: uint32_t

File: src/world/Objects/GameObject.h [hunk 7]
```diff
@@ -267,9 +363,42 @@ struct GameObjectProperties
...
+            uint32 neutralPercent;              // parameter_12
+            uint32 worldstate3;                 // parameter_13
+            uint32 minSuperiority;              // parameter_14
+            uint32 maxSuperiority;              // parameter_15
+            uint32 minTime;                     // parameter_16
+            uint32 maxTime;                     // parameter_17
+            uint32 large;                       // parameter_18
+            uint32 highlight;                   // parameter_19
+            uint32 startingValue;               // parameter_20
+            uint32 unidirectional;              // parameter_21
```
Comment (Appled) [inline src/world/Objects/GameObject.h:400]: uint32_t

File: src/world/Objects/GameObject.h [hunk 8]
```diff
@@ -400,8 +620,89 @@ class SERVER_DECL GameObject : public Object
...
+    {
+        return m_respawnDelayTime == 0 ||
+            (m_respawnTime > 0 && !m_spawnedByDefault) ||
+            (m_respawnTime == 0 && m_spawnedByDefault);
+    }
+
+    bool isSpawnedByDefault() const { return m_spawnedByDefault; }
+    void setSpawnedByDefault(bool b) { m_spawnedByDefault = b; }
+
+    void setSpellId(uint32 id)
```
Comment (Appled) [inline src/world/Objects/GameObject.h:650]: uint32_t

File: src/world/Objects/GameObject.h [hunk 9]
```diff
@@ -587,7 +866,27 @@ class GameObject_Lootable : public GameObject
 
         virtual bool HasLoot() = 0;
 
+        uint16_t getLootMode() const { return m_LootMode; }
+        bool hasLootMode(uint16 lootMode) const { return (m_LootMode & lootMode) != 0; }
+        void setLootMode(uint16 lootMode) { m_LootMode = lootMode; }
+        void addLootMode(uint16 lootMode) { m_LootMode |= lootMode; }
+        void removeLootMode(uint16 lootMode) { m_LootMode &= ~lootMode; }
```
Comment (Appled) [inline src/world/Objects/GameObject.h:873]: uint16_t

File: src/world/Objects/GameObject.h [hunk 10]
```diff
@@ -837,26 +1142,14 @@ class GameObject_FishingNode : public GameObject_Lootable
...
-        void EndFishing(bool abort);
-
-        void EventFishHooked();
-
-        bool IsLootable() { return true; }
-
         bool HasLoot();
 
-    private:
+        bool IsLootable() { return true; }
```
Comment (Appled) [inline src/world/Objects/GameObject.h:1149]: Functions should start with lowercase character

File: src/world/Objects/Transporter.cpp [hunk 11]
```diff
@@ -65,13 +65,29 @@ bool Transporter::Create(uint32_t entry, uint32_t mapid, float x, float y, float
...
 
-    if (!CreateFromProto(entry, mapid, x, y, z, ang))
-        return false;
+    Object::_Create(mapid, x, y, z, ang);
+    setEntry(entry);
+    setLocalRotation(0.0f, 0.0f, 0.0f, 1.0f);
+    setParentRotation(QuaternionData());
+    SetPosition(LocationVector(x, y, z, ang));
+    setDisplayId(gameobject_properties->display_id);
+    setGoType(static_cast<uint8>(gameobject_properties->type));
```
Comment (Appled) [inline src/world/Objects/Transporter.cpp:74]: uint8_t

File: src/world/Objects/Units/Unit.cpp [hunk 12]
```diff
@@ -7755,3 +7755,91 @@ float Unit::getCollisionHeight() const
...
+{
+    if (!gameObj || gameObj->getOwnerGUID() != getGuid())
+        return;
+
+    if (isPlayer())
+        ToPlayer()->setSummonedObject(0);
+
+    gameObj->setOwnerGuid(0);
+
+    for (uint8 i = 0; i < 4; ++i)
```
Comment (Appled) [inline src/world/Objects/Units/Unit.cpp:7787]: uint8_t

File: src/world/Objects/Units/Unit.cpp [hunk 13]
```diff
@@ -7755,3 +7755,91 @@ float Unit::getCollisionHeight() const
...
+    {
+        if (m_ObjectSlots[i] == gameObj->GetUIdFromGUID())
+        {
+            m_ObjectSlots[i] = 0;
+            break;
+        }
+    }
+
+    // GO created by some spell
+    if (uint32 spellid = gameObj->getSpellId())
```
Comment (Appled) [inline src/world/Objects/Units/Unit.cpp:7797]: uint32_t

File: src/world/Objects/Units/Unit.cpp [hunk 14]
```diff
@@ -7755,3 +7755,91 @@ float Unit::getCollisionHeight() const
...
+    m_gameObj.remove(gameObj);
+
+    if (del)
+    {
+        gameObj->setRespawnTime(0);
+        gameObj->Delete();
+    }
+}
+
+void Unit::removeGameObject(uint32 spellid, bool del)
```
Comment (Appled) [inline src/world/Objects/Units/Unit.cpp:7809]: uint32_t

File: src/world/Objects/Units/Unit.cpp [hunk 15]
```diff
@@ -7755,3 +7755,91 @@ float Unit::getCollisionHeight() const
...
+            if (del)
+            {
+                (*i)->setRespawnTime(0);
+                (*i)->Delete();
+            }
+
+            next = m_gameObj.erase(i);
+        }
+        else
+            ++next;
```
Comment (Appled) [inline src/world/Objects/Units/Unit.cpp:7830]: Use braces around else statement if you used them in if statement

File: src/world/Server/Packets/Handlers/LootHandler.cpp [hunk 16]
```diff
@@ -350,216 +356,160 @@ void WorldSession::handleLootReleaseOpcode(WorldPacket& recvPacket)
...
+                if (go->getGoType() == GAMEOBJECT_TYPE_FISHINGHOLE)
+                {                                               // The fishing hole used once more
+                    go->addUse();                               // if the max usage is reached, will be despawned in next tick
+                    if (go->getUseCount() >= dynamic_cast<GameObject_FishingHole*>(go)->getMaxOpen())
+                        go->setLootState(GO_JUST_DEACTIVATED);
+                    else
+                        go->setLootState(GO_READY);
                 }
+                else
+                    go->setLootState(GO_JUST_DEACTIVATED);
```
Comment (Appled) [inline src/world/Server/Packets/Handlers/LootHandler.cpp:410]: Use braces around else statement if you used them in if statement

File: src/world/Spell/SpellEffects.Legacy.cpp [hunk 17]
```diff
@@ -3966,21 +3962,23 @@ void Spell::SpellEffectSummonObject(uint8_t effectIndex)
...
+    WorldMap* map = m_caster->getWorldMap();
     uint32 mapid = u_caster->GetMapId();
     float px = u_caster->GetPositionX();
     float py = u_caster->GetPositionY();
     float pz = u_caster->GetPositionZ();
     float orient = m_caster->GetOrientation();
     float posx = 0, posy = 0, posz = 0;
 
+
+    int32_t duration = ::GetDuration(sSpellDurationStore.LookupEntry(m_spellInfo->getDurationIndex()));
```
Comment (Appled) [inline src/world/Spell/SpellEffects.Legacy.cpp:3974]: Use Spell::getDuration() instead

**Review by Appled — CHANGES_REQUESTED** [review]

File: src/world/Objects/GameObject.h (hunk 10 above)
Comment (Appled) [reply src/world/Objects/GameObject.h:1149]: Then you should define it with override 😉

**Review by Appled — APPROVED** [review]

**Comment by aaron02** [comment]
@Zyres should be rdy and sql updated to match all new sql updates

File: src/world/Chat/Commands/GameObjectCommands.cpp [hunk 18]
```diff
@@ -718,7 +678,7 @@ bool ChatHandler::HandleGOSetFactionCommand(const char* args, WorldSession* m_se
         else
         {
             GreenSystemMessage(m_session, "Faction changed in gameobject_spawns table for spawn ID: %u.", go_spawn->id);
-            WorldDatabase.Execute("UPDATE gameobject_spawns SET faction = %u WHERE id = %u min_build <= %u AND max_build >= %u", go_faction, go_spawn->id, VERSION_STRING, VERSION_STRING);
+            WorldDatabase.Execute("REPLACE INTO gameobject_spawns_overrides VALUES(%u, %u, %u, %3.3lf,%u,%u)", go_spawn->id, VERSION_STRING, VERSION_STRING, gameobject->getScale(), go_faction, gameobject->getFlags());
```
Comment (Zyres) [inline src/world/Chat/Commands/GameObjectCommands.cpp:681]: This is not the same as the "old" UPDATE.....
You try to REPLACE data with id, VERSION, VERSION... but most spawns are multiversion means: id, VERSION-1, VERSION-2.

File: src/world/Chat/Commands/GameObjectCommands.cpp [hunk 19]
```diff
@@ -766,7 +726,7 @@ bool ChatHandler::HandleGOSetFlagsCommand(const char* args, WorldSession* m_sess
         else
         {
             GreenSystemMessage(m_session, "Flags changed in gameobject_spawns table for spawn ID: %u.", go_spawn->id);
-            WorldDatabase.Execute("UPDATE gameobject_spawns SET flags = %u WHERE id = %u AND min_build <= %u AND max_build >= %u", go_flags, go_spawn->id, VERSION_STRING, VERSION_STRING);
+            WorldDatabase.Execute("REPLACE INTO gameobject_spawns_overrides VALUES(%u, %u, %u, %3.3lf,%u,%u)", go_spawn->id, VERSION_STRING, VERSION_STRING, gameobject->getScale(), gameobject->getFactionTemplate(), go_flags);
```
Comment (Zyres) [inline src/world/Chat/Commands/GameObjectCommands.cpp:729]: Same as before, REPLACE is wrong e.g. does not replace anything for multiversion spawns.

File: src/world/Chat/Commands/GameObjectCommands.cpp [hunk 20]
```diff
@@ -920,8 +880,8 @@ bool ChatHandler::HandleGOSetScaleCommand(const char* args, WorldSession* m_sess
         }
         else
         {
-            GreenSystemMessage(m_session, "Scale changed in gameobject_spawns table to %3.3lf for spawn ID: %u.", scale, go_spawn->id);
-            WorldDatabase.Execute("UPDATE gameobject_spawns SET scale = %3.3lf WHERE id = %u AND min_build <= %u AND max_build >= %u", scale, go_spawn->id, VERSION_STRING, VERSION_STRING);
+            GreenSystemMessage(m_session, "Scale changed in gameobject_spawns_overrides table to %3.3lf for spawn ID: %u.", scale, go_spawn->id);
+            WorldDatabase.Execute("REPLACE INTO gameobject_spawns_overrides VALUES(%u, %u, %u, %3.3lf,%u,%u)", go_spawn->id, VERSION_STRING, VERSION_STRING, scale, gameobject->getFactionTemplate(), gameobject->getFlags());
```
Comment (Zyres) [inline src/world/Chat/Commands/GameObjectCommands.cpp:884]: Same....

File: src/world/Objects/GameObject.cpp [hunk 21]
```diff
@@ -254,21 +261,151 @@ void GameObject::setAnimationProgress(uint8_t progress)
...
+{
+    if (m_spawn != nullptr)
+    {
+        WorldDatabase.Execute("DELETE FROM gameobject_spawns WHERE id = %u AND min_build <= %u AND max_build >= %u ", m_spawn->id, VERSION_STRING, VERSION_STRING);
+        WorldDatabase.Execute("DELETE FROM gameobject_spawns_extra WHERE id = %u AND min_build <= %u AND max_build >= %u ", m_spawn->id, VERSION_STRING, VERSION_STRING);
+        WorldDatabase.Execute("DELETE FROM gameobject_spawns_overrides WHERE id = %u AND min_build <= %u AND max_build >= %u ", m_spawn->id, VERSION_STRING, VERSION_STRING);
+    }
+}
+
+void GameObject::saveToDB()
```
Comment (Zyres) [inline src/world/Objects/GameObject.cpp:350]: That function is kind of odd.
If it is a new spawn, save the data with the current Version. If the spawn is already in gameobject_spawns just update the data and keep min/max_build as it was...

**Review by Zyres — CHANGES_REQUESTED** [review]

**Comment by Zyres** [comment]
The wiki should be updated.

**Pull request merged**